ddtrace/appsec/iast/taint_sinks/weak_hash.py

```diff
@@ -0,0 +1,110 @@
+import hashlib
+import sys
+from typing import TYPE_CHECKING
+
+from ddtrace.appsec.iast.constants import EVIDENCE_ALGORITHM_TYPE
+from ddtrace.appsec.iast.constants import VULN_INSECURE_HASHING_TYPE
+from ddtrace.appsec.iast.reporter import report_vulnerability
+from ddtrace.appsec.iast.taint_sinks._base import inject_span
+from ddtrace.internal.logger import get_logger
+from ddtrace.vendor.wrapt import wrap_function_wrapper
```

Comment:
correct, I'd love to see us move to the bytecode patching instead because like Gab says import/patching order matters a lot when doing it the `wrapt` way.

Reply:
In our case we only need these wrapped functions to affect our customers' code, and that code will be the last to load. If a third-party library uses a weak-hashing algorithm it is still a concern, but there would be a CVE about it.

Reply:
> and that code will be the last to load

Being able to remove this assumption is probably for the best IMO. One potential benefit is that, if in the future one would want to dynamically enable/disable some instrumentation, then order must not matter. Any ordering constraints will be an obstacle. Another thing worth thinking about is whether we can also add an import hook on the libraries that are being patched, instead of manually loading them and causing potential side-effects.

Reply:
A note on bytecode patching. We did it at some point in HDiv before moving to AST patching (which for Chapter 0 would not be needed) and getting it right for all cases of function substitution/alteration is almost impossible at the bytecode level, the problem being is that you sometimes have to "backtrack" for example for things like fstrings, function calls in arguments (that could be recursive), classes implementing __call__, and specially (cold sweat) trying to replace things like `+` or `*` or other opcodes for strings for functions. Sometimes that backtracking is not perfect (you can easily interpret it forward but not always so easily backwards unless you keep a full reversible state of the interpreter on memory). For future Python versions there is people working on sort of bytecode plugins (sort of like Rust interpreter plugins) that could serve us at some point, but right now... it gives me PTSD.

For this specific use case of detecting weak hashes however it could work since it doesn't involve most things that we do for IAST, just replacing some functions but I see wrapt as an easier solution. Anyway, since we are installed from sitecustomize: there could really be any case where the modules have been imported before by the client code?

Reply:
> getting it right for all cases of function substitution/alteration is almost impossible at the bytecode level

🤔 Interesting. I wonder why would `wrapt` work in this case, since what we provide in `ddtrace.internal.wrapping` is pretty much the same functionality, except it's done at the bytecode level.

Reply:
I suppose this is from trying to patch the method of an instance? It should work with the actual function from the class.

Reply:
No I didn't, there is below an reproducible example :+1: 
```
import _hashlib
from ddtrace.internal.wrapping import wrap

wrapper = lambda f: f
# 1
method = getattr(_hashlib.HASH, "digest")
wrap(method, wrapper)
>> AttributeError: 'method_descriptor' object has no attribute '__code__'
# 2
wrap(_hashlib.HASH.digest, wrapper)
AttributeError: 'method_descriptor' object has no attribute '__code__'
```

Reply:
> > getting it right for all cases of function substitution/alteration is almost impossible at the bytecode level
> 
> thinking Interesting. I wonder why `wrapt` would work in this case, since what we provide in `ddtrace.internal.wrapping` is pretty much the same functionality, except it's done at the bytecode level.

So I understand the `wrapping.wrap` is for definitions, right? The bytecode-patching-hell I was speaking about was mostly about patching calls (or replacing some inner syntax things with calls) themselves.

Reply:
> So I understand the `wrapping.wrap` is for definitions, right?

Right. All it does is replace the bytecode with that of a function that wraps around the original bytecode, so that extra code can be executed around it, like with `wrapt`. For this to work we need pure Python functions, with an actual bytecode. For functions that call into native we would have to resort to monkey patching unfortunately.